packages/hierarchy-builder/src/hierarchy-builder/internal/operators/LabelGrouping.ts

```diff
@@ -0,0 +1,112 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import naturalCompare from "natural-compare-lite";
+import { from, mergeMap, Observable, tap, toArray } from "rxjs";
+import { Id64 } from "@itwin/core-bentley";
+import { LabelInfo } from "../../EC";
+import { HierarchyNode } from "../../HierarchyNode";
+import { getLogger } from "../../Logging";
+import { createOperatorLoggingNamespace } from "../Common";
+import { sortNodesByLabelOperator } from "./Sorting";
+
+const OPERATOR_NAME = "Grouping.ByLabel";
+/** @internal */
+export const LOGGING_NAMESPACE = createOperatorLoggingNamespace(OPERATOR_NAME);
+
+/** @internal */
+export function createLabelGroupingOperator() {
+  return function (nodes: Observable<HierarchyNode>): Observable<HierarchyNode> {
+    return nodes.pipe(
+      log((n) => `in: ${n.label}`),
+      // need all nodes in one place to group them
+      toArray(),
+      // group all nodes
+      mergeMap((resolvedNodes) => from(createLabelGroupingInformation(resolvedNodes))),
+      // convert intermediate format into a nodes observable
+      mergeMap((groupings) => {
+        const grouped = createGroupingNodes(groupings);
+        const obs = from(grouped);
+        // source observable is expected to stream sorted nodes and we're keeping them in order - only
+        // need to re-sort if we created grouping nodes
+        return grouped.hasLabelGroupingNodes ? obs.pipe(sortNodesByLabelOperator) : obs;
+      }),
+      log((n) => `out: ${n.label}`),
+    );
+  };
+}
+
+interface LabelGroupingInformation {
+  ungrouped: Array<HierarchyNode>;
+  grouped: Map<string, { labelInfo: LabelInfo; groupedNodes: Array<HierarchyNode> }>;
+}
```

Comment:
If we rely on the fact that all input nodes are sorted by label and our new grouping nodes are going to have the same label as grouped ones, then I think we could create a data structure in such a way that doesn't require us to re-sort the nodes afterwards.

Reply:
Added additional property to the "LabelGroupingInformation" called: "order" which tracks in which order the elements have been added to "grouped" or "ungrouped". Then afterwards when creating outNodes I look through this array to keep the same order as it was in the beginning.

Reply:
I think the algorithm could be substantially simplified:

1. Iterate over input nodes. For each node, called _current node_:
2. Check the _last node_ in the output array:
   - If the label is the same as of _current node_:
      - If it's a label grouping node - add _current node_ to _last node's_ children list.
      - Else - replace the _last node_ with a label grouping node that has _last node_ and _current node_ as its children.
   - Else - just add _current node_ to the output array.

The above requires iterating over input array only once, without any intermediate data structures.

Reply:
This is the algorithm that works:
Create output array and assign first node from input nodes to it (if its classGrouping node, then first run this code for its children)

- Iterate over input nodes starting from the second node. For each node, called _current node_:
- Check if the _current node_ is of classGrouping type, if it is, run all this code for the children
- Check the last node in the output array:
  - If the label is the same as of _current node_:
    - If it's a label grouping node:
      - If _current node_ has groupByLabel set to true:
         - add _current node_ to last node's children list.
       - Else:
         - add _current node_ to the output array
         - swap the last two children of the array
     - Else if last node has groupByLabel true:
       - If _current node_ has groupByLabel true:
         - replace the last node with a label grouping node that has last node and _current node_ as its children.
       - Else
         - add _current node_ to the output array
         - swap the last two children of the array
     - Else
       - Add _current node_ to the output array.
  - Else
    - Add _current node_ to the output array.